### ISSUE TITLE
[FR] Lerdge-K/X I²C EEPROM

### ISSUE THREAD

**Issue by Evg33** [MarlinFirmware/Marlin#21313]
How to make I²C EEPROM work?

Chip FM24CL64 F-RAM 64K (8Kx8)
Datasheet: https://www.cypress.com/file/41656/download

Wiring:
![image](https://user-images.githubusercontent.com/7160738/110772560-cbab1d80-826c-11eb-9bd2-b435c894fbe5.png)

A0,A1,A2 on gnd. i.e. EEPROM_DEVICE_ADDRESS 0xA0 ?
![image](https://user-images.githubusercontent.com/7160738/110772528-c1891f00-826c-11eb-9061-d9ef812ce966.png)

```
#define I2C_EEPROM
#define I2C_SDA_PIN                         PG13
#define I2C_SCL_PIN                         PG14
//#define EEPROM_WRITE_DELAY              10
#define EEPROM_DEVICE_ADDRESS           0xA0
//#define MARLIN_EEPROM_SIZE            0x10000 // FM24CL64 F-RAM 64K (8Kx8)
#define MARLIN_EEPROM_SIZE              0x1000  // 4KB for testing
```
In this case M500 gets error "Error writing to EEPROM"

I tried with BL24CXX, but it doesn't compile.
```
#define IIC_BL24CXX_EEPROM            // not compiling
#define IIC_EEPROM_SDA                    PG13
#define IIC_EEPROM_SCL                    PG14
```

What am I doing wrong ?
Need additional code for FM24CL64 ?

**Comment by Evg33** [comment]
p.s.: write address +1

**Comment by GMagician** [comment]
Not sure, I haven't checked code but maybe address is shifted by code, so you may have to define it as 0x50

**Comment by Evg33** [comment]
**GMagician**: yes ! EEPROM_DEVICE_ADDRESS must be 0x50  ! Thank you ! (127 devices on i2c)

But not working :(
```
SENDING:M500
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
....
SENDING:M261 A80 B5
echo:i2c-reply: from:80 bytes:5 data:€€€€€

```

**Comment by Evg33** [comment]
Hmmm. on Lerdge X with same i2c chip it working.... PR with pins later.... After testing...
Maybe I was wrong about the Lerdge-K pins. I need to check again...

**Comment by Evg33** [comment]
Need help....
Maybe need changes in Marlin\buildroot\share\PlatformIO\variants\LERDGE\  ?
for example, here exist pins for Lerdge-X, but not exist for Lerdge-K: 
https://github.com/MarlinFirmware/Marlin/blob/19b94ec3252114ac54bb268bed44679588593729/buildroot/share/PlatformIO/variants/LERDGE/PeripheralPins.c#L106-L128

and TIMERS
https://github.com/MarlinFirmware/Marlin/blob/19b94ec3252114ac54bb268bed44679588593729/buildroot/share/PlatformIO/variants/LERDGE/PeripheralPins.c#L132-L208

**Comment by Evg33** [comment]
and TIMERS
https://github.com/MarlinFirmware/Marlin/blob/19b94ec3252114ac54bb268bed44679588593729/buildroot/share/PlatformIO/variants/LERDGE/PeripheralPins.c#L132-L208

**Comment by Evg33** [comment]
Who can help ?
I can test in HW.

**Comment by GMagician** [comment]
I don't know how your board works, but have you seen #21337, give it a try

**Comment by Evg33** [comment]
I think it won't help. Lerdge-X works with this flash without problems.
But I'll try.

**Comment by Evg33** [comment]
with #21337 problem not solved on Lerdge-K :(

```
SENDING:M502
SENDING:M500
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
echo:Error writing to EEPROM!
...
```

config:
```
#define I2C_EEPROM
#define I2C_SCL_PIN                         PG14
#define I2C_SDA_PIN                         PG13
#define MARLIN_EEPROM_SIZE            0x10000 // FM24CL64 F-RAM 64K (8Kx8)

```

**Comment by dolence** [comment]
Is there anything I can do to help testing this? I'm trying to move from the awful OFW to Marlin on my Lerdge K too.

**Comment by dolence** [comment]
OK, I have this working on K. I 'm not very familiar with git usage and never worked with a project as large as this one, so maybe my approach wasn't the best. I will describe what I did and if anyone else can help me on sending a pull request I would be glad.
The problem is on which pins the EEPROM is wired on K. While X is using pins connected to a real hardware i2c, Lerdge K board have the chip wired to regular pins and probably do some bitbanging. After talking to Jason Nelson, he suggested me to use softi2c. I'm using SlowSoftI2CMaster together with SlowSoftWire, which is a nice wrapper to Wire library.

This is what I did:
On Marlin\platformio.ini I included this line to self resolve the SlowSoftI2CMaster library dependency.
```
# Feature Dependencies
#
[features]
....
SOFT_I2C_EEPROM         = SlowSoftI2CMaster
                          src_filter=+<src/feature/digipot>
```

Then I downloaded SlowSoftWire.cpp and SlowSoftWire.h from [this github](https://github.com/felias-fogg/SlowSoftWirehttps://github.com/felias-fogg/SlowSoftWire) and manually copied it to Marlin\.pio\libdeps\LERDGEK\SlowSoftI2CMaster.

Made this changes to Marlin\src\pins\stm32f4\pins_LERDGE_K.h
```
#define I2C_EEPROM
#define SOFT_I2C_EEPROM //added this to force the use of software i2c
#define I2C_SCL_PIN                         PG14
#define I2C_SDA_PIN                         PG13
#define MARLIN_EEPROM_SIZE               0x10000
```

On Marlin\src\HAL\shared\eeprom_if_i2c.cpp I made this changes:

```
#if ENABLED(I2C_EEPROM) && DISABLED(SOFT_I2C_EEPROM)
  #include <Wire.h> //use Wire library
#elif ENABLED(SOFT_I2C_EEPROM)
  #include <SlowSoftWire.h> //instead, use this softi2c to Wire wrapper
  SlowSoftWire Wire = SlowSoftWire(I2C_SDA_PIN, I2C_SCL_PIN, true);
#endif
```

Before the changes:
![86b3062b-9ee9-4e5b-ad38-af0d077f6d8c](https://user-images.githubusercontent.com/5223003/112674364-97c32100-8e44-11eb-871d-66465c5d3404.jpg)

After the changes, so far working fine!
![6080b139-6d7c-4ead-8eb4-1bf550679399](https://user-images.githubusercontent.com/5223003/112674367-985bb780-8e44-11eb-86b7-b1be813f97c8.jpg)

I was facing some issues with the display initialization. I can confirm this on bugfix and git main. I had to insert two lines from teh standard ST7796S initialization code to make it work properly.

Now I'm having some issues with my dual Z connected to Ex-Motor. At this moment I'm writing some details which I will [update in here](https://github.com/MarlinFirmware/Marlin/issues/21414). Any help is appreciated.

Edit: some typos.

**Comment by GMagician** [comment]
I see that there is some kind of soft i2c support when IIC_BL24CXX_EEPROM is enabled, maybe this can be used this the changed pinout you discovered

**Comment by dolence** [comment]
> I see that there is some kind of soft i2c support when IIC_BL24CXX_EEPROM is enabled, maybe this can be used this the changed pinout you discovered

I will take a look and do some code cleanup this night before I made a PR.

**Comment by dolence** [comment]
If anyone wanna try, I just made a change to my fork which should solve this issue dolence/Marlin@c3940a82218e93e27157501ff35528d643a69411 

Also, I would like to know if someone else is having problems with display initialization on K version 2.0.1 board.

**Comment by sanek88lbl** [comment]
> If anyone wanna try, I just made a change to my fork which should solve this issue [dolence/Marlin@c3940a8](https://github.com/dolence/Marlin/commit/c3940a82218e93e27157501ff35528d643a69411)
> 
> Also, I would like to know if someone else is having problems with display initialization on K version 2.0.1 board.

Hello. I downloaded your Marlin build, checked the work with epprom - it's ok.
I have Lerdge K v2.0.1 and I did not get the display working on your build. I use my last config - it worked. My config is attached.
However, I couldn't start USB OTG, I couldn't even start the build with usb_otg ga VS Code. Does it work for you? How do I do this?

[conf.zip](https://github.com/MarlinFirmware/Marlin/files/6346992/conf.zip)

**Comment by dolence** [comment]
> Hello. I downloaded your Marlin build, checked the work with epprom - it's ok.
> I have Lerdge K v2.0.1 and I did not get the display working on your build. I use my last config - it worked. My config is attached.
> However, I couldn't start USB OTG, I couldn't even start the build with usb_otg ga VS Code. Does it work for you? How do I do this?

Your display definition doesn't look right. If it worked, your colors are probably wrong (RGB channels inverted). Try changing TFT_DRIVER ST7796 to TFT_DRIVER LERDGE_ST7796

Are you sure you are using the [bugfix-2.0.x-LerdgeK](https://github.com/dolence/Marlin/tree/bugfix-2.0.x-LerdgeK) branch?
Unless we have different displays, it should be working.

I will try the USB-OTG thing and see if it works. Actually I'm having a problem with dual Z endstops and can't use my printer until it's figured out and have no idea on how to proceed, so I have decided to abandon this firmware and go back to original until someone else help to solve this #21414

**Comment by sanek88lbl** [comment]
> Your display definition doesn't look right. If it worked, your colors are probably wrong (RGB channels inverted). Try changing TFT_DRIVER ST7796 to TFT_DRIVER LERDGE_ST7796
Are you sure you are using the bugfix-2.0.x-LerdgeK branch?
Unless we have different displays, it should be working.

Yes, I used the wrong branch, but the problem was the display rotated 180 degrees and the touch screen did not work. I copied from my previous build. Now I used your branch with the correct LERDGE driver: the colors did change, but I liked it more with "TFT_DRIVER ST7796"). I also copied the settings of the touchscreen from my previous config-everything worked, thank you!)
[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/6359450/Configuration.zip)


> I will try the USB-OTG thing and see if it works. Actually I'm having a problem with dual Z endstops and can't use my printer until it's figured out and have no idea on how to proceed, so I have decided to abandon this firmware and go back to original until someone else help to solve this #21414

I'll look at your problem with the second Z-axis motor.
In the meantime, I tried building a LERDGEK_USB_FLASH_DRIVE in VS code:

```
Processing LERDGEK_usb_flash_drive (platform: ststm32@~12.0; board: LERDGE; framework: arduino)
-------------------------------------------------------------------------------------------------------------------------------Verbose mode can be enabled via `-v, --verbose` option
KeyError: 'usb-host-msc':
  File "C:\users\user\.platformio\penv\lib\site-packages\platformio\builder\main.py", line 175:
    env.SConscript(item, exports="env")
  File "C:\Users\user\.platformio\packages\tool-scons\scons-local-4.1.0\SCons\Script\SConscript.py", line 591:
    return _SConscript(self.fs, *files, **subst_kw)
  File "C:\Users\user\.platformio\packages\tool-scons\scons-local-4.1.0\SCons\Script\SConscript.py", line 280:
    exec(compile(scriptdata, scriptname, 'exec'), call_stack[-1].globals)
  File "D:\work\3d\firmware\marlin\Marlin-bugfix-2.0.x-LerdgeK\buildroot\share\PlatformIO\scripts\copy_marlin_variant_to_framework.py", line 34:
    FRAMEWORK_DIR = platform.get_package_dir(platform_name)
  File "c:\users\user\.platformio\penv\lib\site-packages\platformio\platform\_packages.py", line 33:
    pkg = self.get_package(name)
  File "c:\users\user\.platformio\penv\lib\site-packages\platformio\platform\_packages.py", line 30:
    return self.pm.get_package(spec or self.get_package_spec(name))
  File "c:\users\user\.platformio\penv\lib\site-packages\platformio\platform\_packages.py", line 22:
    owner=self.packages[name].get("owner"),
================================================= [FAILED] Took 0.84 seconds =================================================

Environment              Status    Duration
-----------------------  --------  ------------
LERDGEK_usb_flash_drive  FAILED    00:00:00.839
============================================ 1 failed, 0 succeeded in 00:00:00.839 ============================================
```

**Comment by dolence** [comment]
> Yes, I used the wrong branch, but the problem was the display rotated 180 degrees and the touch screen did not work. I copied from my previous build. Now I used your branch with the correct LERDGE driver: the colors did change, but I liked it more with "TFT_DRIVER ST7796"). I also copied the settings of the touchscreen from my previous config-everything worked, thank you!)

You should be able to change the display orientation this way:

```
/**
 * TFT Rotation. Set to one of the following values:
 *
 *   TFT_ROTATE_90,  TFT_ROTATE_90_MIRROR_X,  TFT_ROTATE_90_MIRROR_Y,
 *   TFT_ROTATE_180, TFT_ROTATE_180_MIRROR_X, TFT_ROTATE_180_MIRROR_Y,
 *   TFT_ROTATE_270, TFT_ROTATE_270_MIRROR_X, TFT_ROTATE_270_MIRROR_Y,
 *   TFT_MIRROR_X, TFT_MIRROR_Y, TFT_NO_ROTATION
 */
//#define TFT_ROTATION TFT_NO_ROTATION
```

**Comment by Evg33** [comment]
`LERDGEK_usb_flash_drive  FAILED    00:00:00.839`

@sanek88lbl
Use non USB build
https://github.com/MarlinFirmware/Marlin/issues/21302

**Comment by slowbro** [comment]
I think this was implemented in #21812.

**Comment by thisiskeithb** [comment]
> I think this was implemented in #21812.

Yep and EEPROM pins for the Lerdge X were added in #21322.
